addons/mail/static/src/models/emoji_grid_view_row_registry.js

```diff
@@ -43,11 +43,16 @@ registerModel({
             for (let emoji of emojis) {
                 currentItems.push({ emojiOrEmojiInCategory: { emoji } });
                 if (currentItems.length === this.emojiGridViewOwner.amountOfItemsPerRow) {
-                    index++;
                     value.push({ items: currentItems, index });
                     currentItems = [];
+                    index++;
                 }
             }
+            if (currentItems.length > 0) {
+                value.push({ items: currentItems, index });
+                currentItems = [];
+                index++;
+            }
             return value;
```

Comment:
Your fix will do, but I think it would be more efficient and less error-prone to calculate the number of rows beforehand:
```js
computeSearchRows() {
    if (this.emojiGridViewOwner.emojiPickerViewOwner.emojiSearchBarView.currentSearch === "") {
        return clear();
    }
    const emojis = this.messaging.emojiRegistry.allEmojis.filter(this.emojiGridViewOwner._filterEmoji);
    const value = [];
    const numberOfRows = Math.ceil(emojis.length / this.emojiGridViewOwner.amountOfItemsPerRow);
    for (let i = 0; i < numberOfRows; i++) {
        const { amountOfItemsPerRow } = this.emojiGridViewOwner;
        value.push({
            index: i,
            items: emojis.slice(i * amountOfItemsPerRow, i * amountOfItemsPerRow + amountOfItemsPerRow).map(e => ({ emojiOrEmojiInCategory: { emoji: e } })),
        });
    }
    return value;
},
```

Reply:
The problem with this code you suggest is that while it has less LOCs, some lines are too dense

`items: emojis.slice(i * amountOfItemsPerRow, i * amountOfItemsPerRow + amountOfItemsPerRow).map(e => ({ emojiOrEmojiInCategory: { emoji: e } })),` is far too long.

And also you lose the similar shape between `computeSearchRows` and `computeNonSearchRows`.
Both functions do almost the same thing, so it makes sense they should have relatively the same shape.

Maybe we could improve code so they share code better, rather than duplicate shape and some LOCs.
IMO it's better to have code duplication and bad split. And also if the shape are not the same, it makes it harder to adapt both functions at the same time when there is a change that requires to change both computations in a similar fashion.

Reply:
The version of my comment on which you based your reply does not seem to be the latest, since the line you find too long is no longer part of the changes I suggest. The remaining lines that might arguably be too long are non-issues as they can be easily splitted into intermediate variables if necessary.

I agree that the code is better symmetrical, and fortunately these changes should also be applicable to the other compute.

Although this could be an argument in favour of my suggestion, the number of lines of code is not the benefit I wanted you to notice. Calculating the number of rows in advance allows us to flatten the logic by removing the condition inside and outside of the loop. The whole purpose of this compute can be achieved in one single unconditional loop, which reduces the complexity of these lines. I also suspect it to be less error-prone, as it no longer relies on the two elements that led to the issue requiring this PR, i.e. the need to think about when to increment the induction variable and the need to worry about how the row is being filled.

That being said, I wouldn't mind validating your changes as they are, as I don't want to delay a bug fix because of non-functional matters. But I think it would be great to go over this code again at some point. It may be presented as a fix, but it's done in master, so I'm not sure keeping the diff to a minimum as is usually preferred is such a big deal here, meaning I wouldn't find it absurd to take advantage of these changes to do some refactoring. Just tell me what you prefer 🙂

Reply:
Right now it's best to keep symmetry, even though duplicated code is not great. I feel this is the lesser evil over making the functions too different or making a bad function split.